### PR TITLE
Ofir/add context getters setters

### PR DESCRIPTION
**Pull request by itsofirblink**

### CODE REVIEW

File: plugin/action_context.go [hunk 1]
```diff
@@ -34,20 +35,27 @@ func NewActionContext(context map[string]interface{}) *ActionContext {
 }
 
 func (ctx *ActionContext) GetValue(key string) (interface{}, error) {
-	value, ok := ctx.internalContext[key]
-	if !ok {
-		return nil, errors.New(fmt.Sprintf("no entry with name %s found", key))
+	value, err := context.Get(key, ctx.internalContext)
+
```
Comment (danielsblinkops) [inline plugin/action_context.go:39]: Why new line

File: plugin/action_context.go [hunk 2]
```diff
@@ -34,20 +35,27 @@ func NewActionContext(context map[string]interface{}) *ActionContext {
 }
 
 func (ctx *ActionContext) GetValue(key string) (interface{}, error) {
-	value, ok := ctx.internalContext[key]
-	if !ok {
-		return nil, errors.New(fmt.Sprintf("no entry with name %s found", key))
+	value, err := context.Get(key, ctx.internalContext)
```
Comment (danielsblinkops) [inline plugin/action_context.go:38]: Just create a context object which has a .Get() method.
right now its a wierd usage

File: plugin/context/engine.go [hunk 3]
```diff
@@ -0,0 +1,93 @@
+package context
+
+import (
+	"errors"
+	"fmt"
+	"strings"
+)
+
+func getMap(parameter string, context map[string]interface{}) map[string]interface{} {
```
Comment (danielsblinkops) [inline plugin/context/engine.go:9]: What is getMap? What is the map

Comment (danielsblinkops) [inline plugin/context/engine.go:9]: What is parameter, make names more indicative, doc each function

File: plugin/context/engine.go [hunk 4]
```diff
@@ -0,0 +1,93 @@
+package context
+
+import (
+	"errors"
```
Comment (danielsblinkops) [inline plugin/context/engine.go:4]: I think u just make an context class, and have the context map as a member

File: plugin/context/engine.go [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+	"fmt"
+	"strings"
+)
+
+func getMap(parameter string, context map[string]interface{}) map[string]interface{} {
+	// Usage: parameter will be path (json.dot.walking) that the start of the parameter
+	// 		  strings.Split(parameter, ".")[0] -> Should be `inputs` or `parameters`
+	if !strings.Contains(parameter, ".") {
```
Comment (danielsblinkops) [inline plugin/context/engine.go:12]: Dont double scan here, do 
	location := strings.Split(parameter, ".")
	if len(location) ==0 { return nil}
	
Also change location to a more indicative name, line innerObjectKey

File: plugin/context/engine.go [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+
+	location := strings.Split(parameter, ".")[0]
+
+	internalContext, ok := context[location]
+
+	if !ok {
+		return nil
+	}
+
+	return internalContext.(map[string] interface{})
```
Comment (danielsblinkops) [inline plugin/context/engine.go:24]: Do safe casts plsz

File: plugin/context/engine.go [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+}
+
+func getIterator(parameter string, create bool, context map[string]interface{}) (map[string]interface{}, error) {
```
Comment (danielsblinkops) [inline plugin/context/engine.go:27]: You dont return any iterator here... its just a map

Comment (danielsblinkops) [reply plugin/context/engine.go:27]: Also better naming for params, the fuck is create

File: plugin/context/engine.go [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+		return nil, fmt.Errorf("provided parameter is not allowed, must be at-least 1 depth, %v", parameter)
+	}
+
+	iterateMap := getMap(parameter, context)
+	if iterateMap == nil {
+		return nil, fmt.Errorf("failed to get iterator with parameter: %v", parameter)
+	}
+
+	iterator := interface{}(iterateMap)
+	for i := 1; i < depth; i++ {
```
Comment (danielsblinkops) [inline plugin/context/engine.go:44]: Move to outside function which an indiciative name, like resolveInnerItems or something

**Review by danielsblinkops — CHANGES_REQUESTED** [review]

File: plugin/context/engine.go (hunk 3 above)
Comment (itsofirblink) [reply plugin/context/engine.go:9]: Thats the start of the structure, we need to make sure the users can't access something that's not created yet,
we will have `inputs` and `variables` thats the map that you will get access to by `getMap` changed the name to `getInnerContext`

Comment (itsofirblink) [reply plugin/context/engine.go:9]: Changed to `key` and moved the function to the `ActionContext`

File: plugin/context/engine.go (hunk 4 above)
Comment (itsofirblink) [reply plugin/context/engine.go:4]: I'm overriding your functions :)

File: plugin/context/engine.go (hunk 5 above)
Comment (itsofirblink) [reply plugin/context/engine.go:12]: Done :)

File: plugin/context/engine.go (hunk 6 above)
Comment (itsofirblink) [reply plugin/context/engine.go:24]: Will do :)

**Pull request merged**